### PR TITLE
Add API for normalizing addresses

### PR DESCRIPTION
**Pull request by danielocfb**
We want to support symbolization in scenarios where addresses are captured on one system and then symbolized elsewhere (#61). In a nutshell, we are going to perform address "normalization" on the "local" system (the one that captured the addresses), send them to the remote, and then symbolize these normalized addresses on the remote.

This pull request introduces, among other things, the logic for normalizing addresses. Basically, we use the corresponding proc maps entry, parse the contents, identify the ELF files to which those addresses map, and then perform normalization based on the ELF information.

Please see individual commits for descriptions. Note that included here are all but the last commit from  https://github.com/libbpf/blazesym/pull/110, so there is some overlap. I will rebase #110 on top of what is done here (and address the remaining comments).

### CODE REVIEW

File: src/normalize/normalize.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+///
+/// Normalized addresses are reported in the exact same order in which the
+/// non-normalized ones were provided.
+pub fn normalize_user_addrs(addrs: Cow<'_, [Addr]>, pid: u32) -> Result<NormalizedUserAddrs> {
+    let mut addrs = addrs;
+
+    if !util::is_sorted(addrs.iter()) {
+        log::debug!("addresses to normalize not sorted; sorting in line...");
+        let () = addrs.to_mut().sort();
+    }
```
Comment (danielocfb) [inline src/normalize/normalize.rs:141]: I think one thing we can consider doing is having this check done inline with the main logic. That is, just make sure that we see addresses in ascending order. If that is not the case we sort and start over. That would mean we'd still allow for all inputs while not having to pay for an additional pass upfront.

Comment (danielocfb) [reply src/normalize/normalize.rs]: Updated.

File: src/normalize/normalize.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+fn normalize_elf_addr(virt_addr: Addr, entry: &MapsEntry) -> Result<Addr> {
+    let parser = ElfParser::open(&entry.path)?;
+    let phdrs = parser.program_headers()?;
+    let addr = phdrs
+        .iter()
+        .find_map(|phdr| {
+            if phdr.p_type == elf::types::PT_LOAD {
+                if (phdr.p_offset..phdr.p_offset + phdr.p_memsz).contains(&entry.offset) {
+                    let segment_offset = phdr.p_offset - entry.offset;
+                    return Some(virt_addr - entry.range.start - segment_offset as Addr)
```
[the remark this answers is not in the document]
Comment (anakryiko) [reply src/normalize/normalize.rs:103]: 1. use `foff = virt_addr - entry.range.start + entry.offset` to find phdr
2. `foff - phdr.p_offset + phdr.v_offset`

File: src/normalize/normalize.rs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+        let mut prev_addr = addrs.first().copied().unwrap_or_default();
+        // We effectively do a single pass over `addrs`, advancing to the next
+        // proc maps entry whenever the current address is not (or no longer)
+        // contained in the current entry's range.
+        for addr in addrs.iter().copied() {
+            if addr < prev_addr {
+                log::debug!("addresses to normalize not sorted; sorting in line...");
+                let () = addrs.to_mut().sort();
+                continue 'outer
```
[the remark this answers is not in the document]
Comment (d-e-s-o) [reply src/normalize/normalize.rs:169]: Yes, I overlooked that. Good catch. Will rework to require a sorted array for the time being to get this in. We can lift that restriction later and/or add another public API that makes different trade-offs.

File: src/normalize/normalize.rs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            }
+
+            while !entry.range.contains(&addr) {
```
[the remark this answers is not in the document]
Comment (d-e-s-o) [reply src/normalize/normalize.rs:172]: Fair point. Will adjust.

File: src/normalize/normalize.rs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+                entry = entries.next().ok_or_else(|| {
+                    Error::new(
+                        ErrorKind::InvalidInput,
+                        format!("address {addr:x} not found in proc maps for {pid}"),
+                    )
+                })??;
+            }
+
+            let meta_idx = if let Some(meta_idx) = meta_lookup.get(&entry.path) {
```
[the remark this answers is not in the document]
Comment (d-e-s-o) [reply src/normalize/normalize.rs:181]: Sounds good. Yes, we can optimize more. Will probably do that as a follow up, though.

File: src/normalize/normalize.rs [hunk 6]
```diff
@@ -38,6 +43,52 @@ pub struct NormalizedAddrs<M> {
...
+// TODO: Currently look up is always performed based on section name, but there
+//       is also the possibility of iterating notes and checking checking
+//       Elf64_Nhdr.n_type for NT_GNU_BUILD_ID, specifically.
+fn read_build_id(path: &Path) -> Result<Option<String>> {
+    let file = File::open(path)?;
+    let parser = ElfParser::open_file(file)?;
+
+    // The build ID is contained in the `.note.gnu.build-id` section. See
+    // elf(5).
+    if let Ok(idx) = parser.find_section(".note.gnu.build-id") {
```
Comment (anakryiko) [inline src/normalize/normalize.rs:72]: let's at least do sanity checking of note type and name ("GNU")? As you mention in TODO above, kernel and BPF selftests implementations (https://lore.kernel.org/all/20230331093157.1749137-3-jolsa@kernel.org/ and https://github.com/torvalds/linux/blob/master/lib/buildid.c)  don't make an assumption of ".note.gnu.build-id" section (I've been told it's not guaranteed/specified by any starndard, so best not to make assumptions), so maybe let's do it more robustly from the very beginning?

File: src/normalize/normalize.rs [hunk 7]
```diff
@@ -38,6 +43,52 @@ pub struct NormalizedAddrs<M> {
...
+        //         found.
+        let mut bytes = parser.section_data(idx).unwrap();
+        let _header = bytes.read_pod_ref::<BuildIdNote>();
+
+        // Every byte following the header maps to two hex characters.
+        let build_id = bytes
+            .iter()
+            .fold(String::with_capacity(bytes.len() * 2), |mut s, b| {
+                let () = s.push_str(&format!("{b:02x}"));
+                s
```
Comment (anakryiko) [inline src/normalize/normalize.rs:83]: do we need to do string conversion? why not return bytes? it's faster and we avoid questions of lowercase vs uppercase, etc. For simpler testing we can have a helper that converts it to string representation, of course.

File: include/blazesym.h [hunk 8]
```diff
@@ -7,6 +7,20 @@
...
+ */
+typedef enum blaze_user_addr_meta_kind {
+  /**
+   * [`blaze_user_addr_meta_variant::binary`] is valid.
+   */
+  BLAZE_USER_ADDR_KIND_BINARY,
+  /**
+   * [`blaze_user_addr_meta_variant::unknown`] is valid.
+   */
+  BLAZE_USER_ADDR_KIND_UNKNOWN,
```
Comment (anakryiko) [inline include/blazesym.h:21]: let's make sure UNKNOWN value is first

File: src/normalize/normalize.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (anakryiko) [reply src/normalize/normalize.rs:103]: Let's have few more tests upfront? This part is crucial to get right, so having tests that validate that each piece of translation algorithm and formulas are correct is very important. We don't need to have many tests, let's have something similar to urandom_read and urandom_read_lib (just let's make sure we have non-zero file offset in maps file, like I did yesterday locally).

File: src/normalize/normalize.rs [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+/// A type representing normalized user addresses.
+pub type NormalizedUserAddrs = NormalizedAddrs<UserAddrMeta>;
+
+
+/// Normalize a virtual address belonging to an ELF file represented by the
+/// provided [`MapsEntry`].
+fn normalize_elf_addr(virt_addr: Addr, entry: &MapsEntry) -> Result<Addr> {
+    let parser = ElfParser::open(&entry.path)?;
```
[the remark this answers is not in the document]
Comment (d-e-s-o) [reply src/normalize/normalize.rs:95]: As discussed in person, that's not really accurate. We already [strip](https://github.com/libbpf/blazesym/blob/master/src/maps.rs#L99-L103) the `(deleted)` prefix and fall back to using `map_files` then. I agree that we should use `map_files` throughout, but when I tried just stupidly switching over various things blew up and I have not had the time to look into it. So this will get done once I rectify that. But until then, I believe this logic should stay that way, as assumptions here are not different to those made elsewhere in the code.

File: src/normalize/normalize.rs [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+                    _non_exhaustive: (),
+                };
+
+                let meta_idx = normalized.meta.len();
+                let () = normalized.meta.push(UserAddrMeta::Binary(binary));
+                let _ref = meta_lookup.insert(entry.path.to_path_buf(), meta_idx);
+                meta_idx
+            };
+
+            let normalized_addr = normalize_elf_addr(addr, &entry)?;
```
[the remark this answers is not in the document]
Comment (d-e-s-o) [reply src/normalize/normalize.rs:196]: Will work on optimizing this functionality as follow up.

File: src/normalize/normalize.rs (hunk 7 above)
Comment (d-e-s-o) [reply src/normalize/normalize.rs:83]: We don't strictly speaking need it. It felt more appropriate and was what the original design used, but we can postpone conversion. Will just use `Vec`.

**Comment by d-e-s-o** [comment]
Addressed review comments.

File: src/normalize/normalize.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (d-e-s-o) [reply src/normalize/normalize.rs]: I can add one more test to force a non-zero file offset in proc maps, yes.

[the remark this answers is not in the document]
Comment (danielocfb) [reply src/normalize/normalize.rs]: Done now. Please re-review.

File: src/c_api/normalize.rs [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+                .as_mut_ptr()
+            },
+            addr_count: other.addrs.len(),
+            addrs: unsafe {
+                Box::into_raw(
+                    other
+                        .addrs
+                        .into_iter()
+                        .map(blaze_normalized_addr::from)
+                        .collect::<Vec<_>>()
```
Comment (anakryiko) [inline src/c_api/normalize.rs:258]: will this approach pre-allocate the right size of Vec<> upfront, or it will be gradually resizing it. Asking because we do know the final size of the Vec, so it's more efficient to presize it

File: src/normalize/normalize.rs [hunk 12]
```diff
@@ -0,0 +1,421 @@
...
+fn normalize_elf_addr(virt_addr: Addr, entry: &MapsEntry) -> Result<Addr> {
+    let file_off = virt_addr - entry.range.start + entry.offset as usize;
+    let parser = ElfParser::open(&entry.path)?;
+    let phdrs = parser.program_headers()?;
+    let addr = phdrs
+        .iter()
+        .find_map(|phdr| {
+            if phdr.p_type == elf::types::PT_LOAD {
+                if (phdr.p_offset..phdr.p_offset + phdr.p_memsz).contains(&(file_off as u64)) {
+                    return Some(file_off + phdr.p_vaddr as usize - phdr.p_offset as usize)
```
Comment (anakryiko) [inline src/normalize/normalize.rs:122]: nit: to avoid (somewhat hypothetical)  overflow, it's best to subtract p_offset before adding p_vaddr

**Review by anakryiko — APPROVED** [review]
besides the build_id in Unknown (and few nits), lgtm

File: src/c_api/normalize.rs [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+
+
+/// C compatible version of [`Unknown`].
+#[allow(non_camel_case_types)]
+#[derive(Debug)]
+pub struct blaze_user_addr_meta_unknown {
+    /// The length of the build ID, in bytes.
+    build_id_len: usize,
+    /// The optional build ID of the binary, if found.
+    build_id: *mut u8,
```
[the remark this answers is not in the document]
Comment (danielocfb) [reply src/c_api/normalize.rs:124]: I'll remove it.

File: src/normalize/normalize.rs (hunk 12 above)
Comment (danielocfb) [reply src/normalize/normalize.rs:122]: Sure, done.

File: src/c_api/normalize.rs (hunk 11 above)
Comment (danielocfb) [reply src/c_api/normalize.rs:258]: As discussed in person, yeah, the iterator knows it size and collect will take that into account and presize accordingly in advance.
